src/radar_control.py

```diff
@@ -681,6 +718,76 @@ def run_cfs_scan(
     return freq_data
 
 
+def cfs_block(ave_params, cfs_params_dict, cfs_sockets):
+    aveperiod = ave_params.aveperiod
+    if (
+        cfs_params_dict[aveperiod].last_cfs_set_time[aveperiod.beam_iter]
+        < time.time() - ave_params.aveperiod.cfs_stable_time
+        or aveperiod.cfs_always_run
+    ):
+        # Only let CFS run after the user set stable time has
+        # passed to prevent CFS from switching freqs to quickly
+        ave_params.sequence = aveperiod.cfs_sequence
+        ave_params.decimation_scheme = aveperiod.cfs_sequence.decimation_scheme
+
+        freq_spectrum = run_cfs_scan(
+            cfs_sockets[0],
+            cfs_sockets[1],
+            ave_params,
+            cfs_sockets[2],
+        )
+        ave_params.num_sequences += 1
+        ave_params.cfs_scan_flag = False
+        cfs_params_dict[aveperiod].cfs_freq = freq_spectrum.cfs_freq
+
+        for ind, dset in enumerate(freq_spectrum.output_datasets):
+            cfs_params_dict[aveperiod].cfs_mags[aveperiod.cfs_slice_ids[ind]] = (
+                dset.cfs_data
+            )
+
+        if (
+            cfs_params_dict[aveperiod].last_cfs_set_time[aveperiod.beam_iter]
+            < time.time() - ave_params.aveperiod.cfs_stable_time
+        ):
+            cfs_params_dict[aveperiod].last_cfs_set_time[aveperiod.beam_iter] = (
+                time.time()
+            )
+
+            if (
+                not cfs_params_dict[aveperiod].set_new_freq[aveperiod.beam_iter]
+                and aveperiod.cfs_pwr_threshold is not None
+            ):
+                # If using a user set power threshold to trigger CFS freq setting, check
+                # if any power related condition are triggered
+                cfs_params_dict[aveperiod].check_update_freq(
+                    freq_spectrum,
+                    aveperiod.cfs_slice_ids,
+                    aveperiod.cfs_pwr_threshold,
+                    aveperiod.beam_iter,
+                )
+
+            if (
+                cfs_params_dict[aveperiod].set_new_freq[aveperiod.beam_iter]
+                or aveperiod.cfs_pwr_threshold is None
+            ):
```

Comment:
de-nest

Reply:
Personally I think this one would be fine to leave, but I'll de-nest it